File: server/invoice_pdf.py
"""Generate branded invoice PDFs for Rinse & Rise Laundryrite."""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

from fpdf import FPDF

IST = ZoneInfo("Asia/Kolkata")

ROOT = Path(__file__).resolve().parent.parent
LOGO_PATH = ROOT / "assets" / "logo.png"
REVIEW_QR_PATH = ROOT / "assets" / "google-review-qr.jpg"
INVOICE_DIR = ROOT / "data" / "invoices"

GOOGLE_REVIEW_URL = (
    "https://google.com/maps/place//data=!4m3!3m2!1s0x3bae6d01fa8c4225:0x865cd70a4ba3ada4"
    "!12e1?source=g.page.m.ia._&laa=nmx-review-solicitation-ia2"
)

ORANGE = (242, 101, 34)
ORANGE_DARK = (217, 84, 15)
BLACK = (26, 26, 26)
GRAY = (100, 100, 100)
LIGHT = (245, 245, 245)
WHITE = (255, 255, 255)
BORDER = (220, 220, 220)

MARGIN = 14
CONTENT_W = 182
HEADER_H = 44
FOOTER_H = 46
QR_SIZE = 22
LOGO_SIZE = 22

TABLE_COLS = [
    ("#", 10),
    ("Item", 58),
    ("Service", 36),
    ("Qty", 22),
    ("Rate", 26),
    ("Amount", 30),
]

PAYMENT_TYPE_LABELS = {"cash": "Cash", "upi": "UPI Online"}
PAYMENT_INFO_LABELS = {"pre-payment": "Pre Payment", "post-payment": "Post Payment"}

SHOP = {
    "phone": "9591506548",
    "hours": "9 AM - 9 PM",
    "address_lines": (
        "Shop No 1, Site No 211, 15th Cross, 4th Main,",
        "Ananth Nagar Phase 1, Electronic City, Bangalore 560100",
    ),
    "tagline": "Free Pickup & Delivery  |  Express 24-Hr Delivery",
}

DELIVERY_TIME_LABELS = {
    "morning": "Morning",
    "afternoon": "Afternoon",
    "evening": "Evening",
}


def _normalize_delivery_date_text(text: str) -> str:
    text = _strip_delivery_time_suffix((text or "").strip())
    if not text:
        return "-"
    for fmt in ("%a, %d %b, %Y", "%d %b, %Y", "%d %b %Y"):
        try:
            return datetime.strptime(text, fmt).strftime("%d %b %Y")
        except ValueError:
            continue
    return text


def format_delivery_schedule(bill: dict[str, Any]) -> str:
    date_val = (bill.get("deliveryDate") or "").strip()
    if date_val:
        try:
            year, month, day = map(int, date_val.split("-"))
            dt = datetime(year, month, day)
            return dt.strftime("%d %b %Y")
        except ValueError:
            return date_val

    display = (bill.get("deliveryDisplay") or "").strip()
    if " · " in display:
        display = display.split(" · ", 1)[1].strip()
    return _normalize_delivery_date_text(display)


def _strip_delivery_time_suffix(text: str) -> str:
    text = (text or "").strip()
    if not text:
        return ""
    for label in DELIVERY_TIME_LABELS.values():
        suffix = f", {label}"
        if text.endswith(suffix):
            return text[: -len(suffix)].strip()
    if "," in text:
        last = text.rsplit(",", 1)[-1].strip().lower()
        if last in DELIVERY_TIME_LABELS or last in {v.lower() for v in DELIVERY_TIME_LABELS.values()}:
            return text.rsplit(",", 1)[0].strip()
    return text


def format_service_name(name: str) -> str:
    cleaned = (name or "").strip()
    if cleaned.lower() in ("lundry", "laundry"):
        return "Laundry"
    return cleaned or "-"


def format_payment_summary(bill: dict[str, Any]) -> str:
    parts: list[str] = []
    pt = PAYMENT_TYPE_LABELS.get((bill.get("paymentType") or "").strip(), "")
    pi = PAYMENT_INFO_LABELS.get((bill.get("paymentInfo") or "").strip(), "")
    if pt:
        parts.append(pt)
    if pi:
        parts.append(pi)
    return " · ".join(parts) if parts else "-"


def format_qty_display(item: dict[str, Any]) -> str:
    qty = float(item.get("qty") or 0)
    if _is_kg_item(item):
        text = str(int(qty)) if qty == int(qty) else f"{qty:.1f}"
        return f"{text} kg"
    return f"{int(qty)} pc"


def _is_kg_item(item: dict[str, Any]) -> bool:
    if item.get("unit") == "kg":
        return True
    if item.get("unit") == "pc":
        return False
    name = (item.get("name") or "").lower()
    return "/kg" in name or name in {
        "wash and fold 80/kg",
        "wash and iron 125/kg",
        "premium laundry 200/kg",
    }


def format_currency(amount: float | int) -> str:
    return f"Rs. {float(amount or 0):,.0f}"


def format_qty_rate_line(item: dict[str, Any]) -> str:
    qty = float(item.get("qty") or 0)
    rate = format_currency(item.get("rate", 0))
    if _is_kg_item(item):
        text = str(int(qty)) if qty == int(qty) else f"{qty:.1f}"
        return f"{text} kg × {rate}/kg"
    return f"{int(qty)} pc × {rate}"


def format_bill_date(value: str | None) -> str:
    if not value:
        dt = datetime.now(IST)
    else:
        try:
            raw = str(value).replace("Z", "+00:00")
            dt = datetime.fromisoformat(raw)
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            dt = dt.astimezone(IST)
        except ValueError:
            return str(value)
    return dt.strftime("%d %b %Y, %I:%M %p").replace("AM", "am").replace("PM", "pm")


def build_whatsapp_message(bill: dict[str, Any]) -> str:
    bill = normalize_bill_for_pdf(bill)
    name = (bill.get("customerName") or "Customer").strip()
    delivery = format_delivery_schedule(bill)
    total = format_currency(bill.get("total", 0))
    bill_no = bill.get("billNo", "")
    subtotal = float(bill.get("subtotal") or 0)
    discount = float(bill.get("discountAmount") or 0)
    discount_pct = float(bill.get("discountPercent") or 0)

    lines = [
        f"Hi {name},",
        "",
        "Your invoice from *Rinse & Rise Laundryrite* is attached.",
        "",
        f"Bill No: *#{bill_no}*",
        f"Delivery: {delivery}",
        "",
        "*Items:*",
    ]

    for i, item in enumerate(bill.get("items") or [], 1):
        amount = float(item.get("rate") or 0) * float(item.get("qty") or 0)
        lines.append(f"{i}. {item.get('name', '-')}")
        lines.append(f"   {format_qty_rate_line(item)} = *{format_currency(amount)}*")

    lines.append("")
    if discount > 0:
        lines.append(f"Subtotal: {format_currency(subtotal)}")
        lines.append(f"Discount ({discount_pct:.0f}%): - {format_currency(discount)}")
    lines.append(f"*TOTAL: {total}*")
    lines.extend(
        [
            "",
            "Thank you for choosing us!",
            "Rinse - Rise - Repeat",
            f"Call: {SHOP['phone']} | {SHOP['hours']}",
            "",
            "*Leave us a Google Review:*",
            GOOGLE_REVIEW_URL,
        ]
    )
    return "\n".join(lines)


def invoice_filename(bill: dict[str, Any]) -> str:
    bill_no = str(bill.get("billNo") or bill.get("id") or "bill")
    safe = "".join(c if c.isalnum() or c in "-_" else "-" for c in bill_no)
    return f"RinseRise-Invoice-{safe}.pdf"


class InvoicePDF(FPDF):
    def footer(self):
        footer_top = self.h - FOOTER_H
        qr_block_w = QR_SIZE + 10 if REVIEW_QR_PATH.is_file() else 0
        text_w = self.w - (MARGIN * 2) - qr_block_w
        text_center_x = MARGIN + text_w / 2

        self.set_draw_color(*BORDER)
        self.set_line_width(0.2)
        self.line(MARGIN, footer_top, self.w - MARGIN, footer_top)

        line_y = footer_top + 6
        self.set_font("Helvetica", "B", 9)
        self.set_text_color(*BLACK)
        self.set_xy(MARGIN, line_y)
        self.cell(text_w, 5, "Thank you for choosing Rinse & Rise Laundryrite!", align="C")

        line_y += 6
        self.set_font("Helvetica", "", 8)
        self.set_text_color(*GRAY)
        self.set_xy(MARGIN, line_y)
        self.cell(text_w, 4, SHOP["tagline"], align="C")

        line_y += 5
        self.set_xy(MARGIN, line_y)
        self.cell(text_w, 4, f"Call: {SHOP['phone']}  |  {SHOP['hours']}", align="C")

        line_y += 5
        for addr_line in SHOP["address_lines"]:
            self.set_xy(MARGIN, line_y)
            self.cell(text_w, 4, addr_line, align="C")
            line_y += 4

        if REVIEW_QR_PATH.is_file():
            qr_x = self.w - MARGIN - QR_SIZE
            qr_y = footer_top + (FOOTER_H - QR_SIZE - 8) / 2
            self.image(str(REVIEW_QR_PATH), x=qr_x, y=qr_y, w=QR_SIZE, h=QR_SIZE)
            self.set_font("Helvetica", "B", 7)
            self.set_text_color(*ORANGE)
            self.set_xy(qr_x - 4, qr_y + QR_SIZE + 1)
            self.cell(QR_SIZE + 8, 3, "Rate us on Google", align="C")
            self.set_font("Helvetica", "", 6)
            self.set_text_color(*GRAY)
            self.set_xy(qr_x - 4, qr_y + QR_SIZE + 4.5)
            self.cell(QR_SIZE + 8, 3, "Scan to review", align="C")


def _draw_header(pdf: InvoicePDF, bill: dict[str, Any]) -> None:
    page_w = pdf.w
    pdf.set_fill_color(*ORANGE)
    pdf.rect(0, 0, page_w, HEADER_H, "F")
    pdf.set_fill_color(*ORANGE_DARK)
    pdf.rect(0, HEADER_H - 2, page_w, 2, "F")

    logo_y = (HEADER_H - LOGO_SIZE) / 2
    if LOGO_PATH.is_file():
        pdf.image(str(LOGO_PATH), x=MARGIN, y=logo_y, w=LOGO_SIZE, h=LOGO_SIZE)

    text_x = MARGIN + LOGO_SIZE + 6
    brand_lines = [
        ("Helvetica", "B", 19, "RINSE & RISE", 8),
        ("Helvetica", "B", 10, "LAUNDRYRITE", 6),
        ("Helvetica", "", 8, "Rinse · Rise · Repeat", 5),
    ]
    block_h = sum(h for *_, h in brand_lines)
    brand_y = (HEADER_H - block_h) / 2
    pdf.set_text_color(*WHITE)
    for font, style, size, text, lh in brand_lines:
        pdf.set_font(font, style, size)
        pdf.set_xy(text_x, brand_y)
        pdf.cell(90, lh, text)
        brand_y += lh

    invoice_lines = [
        ("Helvetica", "B", 24, "INVOICE", 9),
        ("Helvetica", "B", 11, f"#{bill.get('billNo', '-')}", 6),
        ("Helvetica", "", 9, format_bill_date(bill.get("createdAt")), 5),
    ]
    block_h = sum(h for *_, h in invoice_lines)
    inv_y = (HEADER_H - block_h) / 2
    for font, style, size, text, lh in invoice_lines:
        pdf.set_font(font, style, size)
        pdf.set_xy(MARGIN, inv_y)
        pdf.cell(page_w - MARGIN * 2, lh, text, align="R")
        inv_y += lh


def _draw_info_box(
    pdf: InvoicePDF,
    x: float,
    y: float,
    w: float,
    h: float,
    title: str,
    primary: str,
    secondary: str = "",
) -> None:
    pdf.set_fill_color(*LIGHT)
    pdf.set_draw_color(*BORDER)
    pdf.rect(x, y, w, h, "FD")

    pdf.set_text_color(*ORANGE)
    pdf.set_font("Helvetica", "B", 7.5)
    pdf.set_xy(x + 5, y + 5)
    pdf.cell(w - 10, 4, title.upper())

    pdf.set_text_color(*BLACK)
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_xy(x + 5, y + 12)
    pdf.cell(w - 10, 5, primary or "-")

    if secondary:
        pdf.set_font("Helvetica", "", 9)
        pdf.set_text_color(*GRAY)
        pdf.set_xy(x + 5, y + 19)
        pdf.cell(w - 10, 5, secondary)


def _table_right_edge() -> float:
    return MARGIN + CONTENT_W


def _col_x(col_index: int) -> float:
    x = MARGIN
    for i in range(col_index):
        x += TABLE_COLS[i][1]
    return x


def _draw_table_row(
    pdf: InvoicePDF,
    y: float,
    values: list[str],
    *,
    header: bool = False,
    fill: tuple[int, int, int] | None = None,
    bold_last: bool = False,
) -> float:
    row_h = 9 if header else 8
    aligns = ("C", "L", "L", "C", "R", "R")

    if header:
        pdf.set_fill_color(*ORANGE)
        pdf.set_text_color(*WHITE)
        pdf.set_font("Helvetica", "B", 8.5)
    else:
        pdf.set_fill_color(*(fill or WHITE))
        pdf.set_text_color(*BLACK)
        pdf.set_font("Helvetica", "", 9)

    for i, ((_, width), val, align) in enumerate(zip(TABLE_COLS, values, aligns)):
        x = _col_x(i)
        pdf.set_draw_color(*BORDER)
        pdf.set_xy(x, y)
        if bold_last and i == len(values) - 1:
            pdf.set_font("Helvetica", "B", 9)
        pdf.cell(width, row_h, val, border=1, fill=True, align=align)

    return y + row_h


def _draw_totals_block(
    pdf: InvoicePDF,
    y: float,
    *,
    subtotal: float,
    discount: float,
    discount_pct: float,
    total: float,
) -> float:
    amount_w = TABLE_COLS[-1][1]
    label_w = TABLE_COLS[-2][1]
    block_w = label_w + amount_w
    block_x = _table_right_edge() - block_w
    amount_x = block_x + label_w
    line_h = 7

    if discount > 0:
        pdf.set_font("Helvetica", "", 9)
        pdf.set_text_color(*GRAY)
        pdf.set_xy(block_x, y)
        pdf.cell(label_w, line_h, "Subtotal", align="R")
        pdf.set_xy(amount_x, y)
        pdf.cell(amount_w, line_h, format_currency(subtotal), align="R")
        y += line_h

        pdf.set_text_color(*ORANGE)
        pdf.set_xy(block_x, y)
        pdf.cell(label_w, line_h, f"Discount ({discount_pct:.0f}%)", align="R")
        pdf.set_xy(amount_x, y)
        pdf.cell(amount_w, line_h, f"- {format_currency(discount)}", align="R")
        y += line_h + 2

    total_h = 11
    pdf.set_fill_color(*ORANGE)
    pdf.rect(block_x, y, block_w, total_h, "F")
    pdf.set_text_color(*WHITE)
    pdf.set_font("Helvetica", "B", 11)
    pdf.set_xy(block_x + 4, y + 3)
    pdf.cell(label_w - 4, 6, "TOTAL")
    pdf.set_xy(amount_x, y + 3)
    pdf.cell(amount_w - 2, 6, format_currency(total), align="R")
    return y + total_h


def normalize_bill_for_pdf(bill: dict[str, Any]) -> dict[str, Any]:
    normalized = dict(bill)
    items = []
    for raw in bill.get("items") or []:
        item = dict(raw)
        item["qty"] = float(item.get("qty") or 0)
        if not item.get("unit"):
            item["unit"] = "kg" if _is_kg_item(item) else "pc"
        items.append(item)
    normalized["items"] = items
    return normalized


def generate_invoice_pdf(bill: dict[str, Any], output_path: Path | str | None = None) -> Path:
    bill = normalize_bill_for_pdf(bill)
    INVOICE_DIR.mkdir(parents=True, exist_ok=True)
    bill_id = bill.get("id") or "draft"
    path = Path(output_path) if output_path else INVOICE_DIR / f"bill-{bill_id}.pdf"

    pdf = InvoicePDF("P", "mm", "A4")
    pdf.set_auto_page_break(auto=True, margin=FOOTER_H + 8)
    pdf.set_margins(MARGIN, MARGIN, MARGIN)
    pdf.add_page()

    _draw_header(pdf, bill)

    y = HEADER_H + 8
    gap = 5
    box_w = (CONTENT_W - gap * 2) / 3
    box_h = 28

    _draw_info_box(
        pdf,
        MARGIN,
        y,
        box_w,
        box_h,
        "Bill To",
        (bill.get("customerName") or "Customer").strip(),
        (bill.get("customerPhone") or "-").strip(),
    )
    _draw_info_box(
        pdf,
        MARGIN + box_w + gap,
        y,
        box_w,
        box_h,
        "Delivery",
        format_delivery_schedule(bill),
    )
    _draw_info_box(
        pdf,
        MARGIN + (box_w + gap) * 2,
        y,
        box_w,
        box_h,
        "Payment",
        format_payment_summary(bill),
    )

    y += box_h + 10

    header_vals = [label for label, _ in TABLE_COLS]
    y = _draw_table_row(pdf, y, header_vals, header=True)

    items = bill.get("items") or []
    for i, item in enumerate(items):
        rate = float(item.get("rate") or 0)
        qty = float(item.get("qty") or 0)
        amount = rate * qty
        fill = LIGHT if i % 2 == 0 else WHITE
        row = [
            str(i + 1),
            str(item.get("name") or "-")[:48],
            format_service_name(str(item.get("service") or "-"))[:28],
            format_qty_display(item),
            format_currency(rate) + ("/kg" if _is_kg_item(item) else ""),
            format_currency(amount),
        ]
        y = _draw_table_row(pdf, y, row, fill=fill, bold_last=True)

    y += 6
    subtotal = float(bill.get("subtotal") or 0)
    discount = float(bill.get("discountAmount") or 0)
    discount_pct = float(bill.get("discountPercent") or 0)
    total = float(bill.get("total") or 0)

    _draw_totals_block(
        pdf,
        y,
        subtotal=subtotal,
        discount=discount,
        discount_pct=discount_pct,
        total=total,
    )

    pdf.output(str(path))
    return path

File: js/invoice-pdf.js
/**
 * Branded A4 invoice PDF for Rinse & Rise Laundryrite.
 * Requires jsPDF + jspdf-autotable (loaded in index.html).
 */
const InvoicePdf = (() => {
  const ORANGE = [242, 101, 34];
  const ORANGE_DARK = [217, 84, 15];
  const BLACK = [26, 26, 26];
  const GRAY = [100, 100, 100];
  const LIGHT = [245, 245, 245];
  const BORDER = [220, 220, 220];

  const SHOP = {
    phone: "9591506548",
    hours: "9 AM – 9 PM",
    addressLines: [
      "Shop No 1, Site No 211, 15th Cross, 4th Main,",
      "Ananth Nagar Phase 1, Electronic City, Bangalore 560100",
    ],
    tagline: "Free Pickup & Delivery  |  Express 24-Hr Delivery",
  };

  const PAYMENT_TYPE_LABELS = { cash: "Cash", upi: "UPI Online" };
  const PAYMENT_INFO_LABELS = { "pre-payment": "Pre Payment", "post-payment": "Post Payment" };

  const MARGIN = 14;
  const CONTENT_W = 182;
  const HEADER_H = 44;
  const FOOTER_H = 46;
  const QR_SIZE = 22;
  const LOGO_SIZE = 22;
  const TABLE_COLS = [10, 58, 36, 22, 26, 30];

  function formatCurrency(amount) {
    return "Rs. " + Number(amount || 0).toLocaleString("en-IN");
  }

  const KG_ITEM_NAMES = new Set([
    "Wash and Fold 80/kg",
    "Wash and Iron 125/kg",
    "Premium Laundry 200/kg",
  ]);

  function isKgItem(item) {
    if (!item) return false;
    if (item.unit === "kg") return true;
    if (item.unit === "pc") return false;
    const name = item.name || "";
    return KG_ITEM_NAMES.has(name) || /\/kg/i.test(name);
  }

  function formatQtyDisplay(item) {
    const qty = Number(item.qty) || 0;
    if (isKgItem(item)) {
      const text = qty % 1 === 0 ? String(qty) : qty.toFixed(1);
      return `${text} kg`;
    }
    return `${qty} pc`;
  }

  const DELIVERY_TIME_LABELS = {
    morning: "Morning",
    afternoon: "Afternoon",
    evening: "Evening",
  };

  function formatServiceName(name) {
    const cleaned = String(name || "").trim();
    if (cleaned.toLowerCase() === "lundry" || cleaned.toLowerCase() === "laundry") return "Laundry";
    return cleaned || "—";
  }

  function formatPaymentSummary(bill) {
    const parts = [];
    const pt = PAYMENT_TYPE_LABELS[bill.paymentType] || "";
    const pi = PAYMENT_INFO_LABELS[bill.paymentInfo] || "";
    if (pt) parts.push(pt);
    if (pi) parts.push(pi);
    return parts.length ? parts.join(" · ") : "—";
  }

  function stripDeliveryTimeSuffix(text) {
    const value = String(text || "").trim();
    if (!value) return "";
    for (const label of Object.values(DELIVERY_TIME_LABELS)) {
      if (value.endsWith(`, ${label}`)) return value.slice(0, -(`, ${label}`.length)).trim();
    }
    return value;
  }

  function formatDeliverySchedule(bill) {
    const dateVal = (bill.deliveryDate || "").trim();
    if (dateVal) {
      const [year, month, day] = dateVal.split("-").map(Number);
      const delivery = new Date(year, month - 1, day);
      if (!Number.isNaN(delivery.getTime())) {
        return delivery.toLocaleDateString("en-IN", {
          day: "2-digit",
          month: "short",
          year: "numeric",
        });
      }
      return dateVal;
    }

    let display = (bill.deliveryDisplay || "").trim();
    if (display.includes(" · ")) display = display.split(" · ").slice(1).join(" · ").trim();
    return stripDeliveryTimeSuffix(display) || "—";
  }

  function formatBillDate(value) {
    const d = value instanceof Date ? value : new Date(value || Date.now());
    if (Number.isNaN(d.getTime())) return "—";
    return d
      .toLocaleString("en-IN", {
        timeZone: "Asia/Kolkata",
        day: "2-digit",
        month: "short",
        year: "numeric",
        hour: "2-digit",
        minute: "2-digit",
        hour12: true,
      })
      .replace(/\s?(AM|PM)/i, (_, meridiem) => ` ${meridiem.toLowerCase()}`);
  }

  let logoDataUrl = null;
  let reviewQrDataUrl = null;

  function drawInfoBox(doc, x, y, w, h, title, primary, secondary) {
    doc.setFillColor(...LIGHT);
    doc.setDrawColor(...BORDER);
    doc.rect(x, y, w, h, "FD");
    doc.setTextColor(...ORANGE);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(7.5);
    doc.text(title.toUpperCase(), x + 5, y + 8);
    doc.setTextColor(...BLACK);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(10);
    doc.text(primary || "—", x + 5, y + 15);
    if (secondary) {
      doc.setFont("helvetica", "normal");
      doc.setFontSize(9);
      doc.setTextColor(...GRAY);
      doc.text(secondary, x + 5, y + 22);
    }
  }

  function drawHeader(doc, bill, pageW, margin, logo) {
    doc.setFillColor(...ORANGE);
    doc.rect(0, 0, pageW, HEADER_H, "F");
    doc.setFillColor(...ORANGE_DARK);
    doc.rect(0, HEADER_H - 2, pageW, 2, "F");

    const logoY = (HEADER_H - LOGO_SIZE) / 2;
    doc.addImage(logo, "PNG", margin, logoY, LOGO_SIZE, LOGO_SIZE);

    const textX = margin + LOGO_SIZE + 6;
    const brandBlockH = 19;
    let brandY = (HEADER_H - brandBlockH) / 2;
    doc.setTextColor(255, 255, 255);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(19);
    doc.text("RINSE & RISE", textX, brandY + 6);
    doc.setFontSize(10);
    doc.text("LAUNDRYRITE", textX, brandY + 12);
    doc.setFont("helvetica", "normal");
    doc.setFontSize(8);
    doc.text("Rinse \u00B7 Rise \u00B7 Repeat", textX, brandY + 17);

    const invBlockH = 20;
    let invY = (HEADER_H - invBlockH) / 2;
    doc.setFont("helvetica", "bold");
    doc.setFontSize(24);
    doc.text("INVOICE", pageW - margin, invY + 8, { align: "right" });
    doc.setFontSize(11);
    doc.text(`#${bill.billNo || "—"}`, pageW - margin, invY + 14, { align: "right" });
    doc.setFont("helvetica", "normal");
    doc.setFontSize(9);
    doc.text(formatBillDate(bill.createdAt), pageW - margin, invY + 19, { align: "right" });
  }

  async function loadReviewQr() {
    if (reviewQrDataUrl) return reviewQrDataUrl;
    const res = await fetch("assets/google-review-qr.jpg");
    if (!res.ok) return null;
    const blob = await res.blob();
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        reviewQrDataUrl = reader.result;
        resolve(reviewQrDataUrl);
      };
      reader.onerror = reject;
      reader.readAsDataURL(blob);
    });
  }

  async function loadLogo() {
    if (logoDataUrl) return logoDataUrl;
    const res = await fetch("assets/logo.png");
    if (!res.ok) throw new Error("Could not load shop logo.");
    const blob = await res.blob();
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        logoDataUrl = reader.result;
        resolve(logoDataUrl);
      };
      reader.onerror = reject;
      reader.readAsDataURL(blob);
    });
  }

  function normalizeBillForPdf(bill) {
    return {
      ...bill,
      items: (bill.items || []).map((raw) => {
        const item = { ...raw, qty: Number(raw.qty) || 0 };
        if (!item.unit) item.unit = isKgItem(item) ? "kg" : "pc";
        return item;
      }),
    };
  }

  async function generate(bill) {
    bill = normalizeBillForPdf(bill);
    if (!window.jspdf?.jsPDF) {
      throw new Error("PDF library not loaded. Check your internet connection and refresh.");
    }

    const { jsPDF } = window.jspdf;
    const doc = new jsPDF({ unit: "mm", format: "a4", orientation: "portrait" });
    const margin = MARGIN;
    const pageW = doc.internal.pageSize.getWidth();
    const pageH = doc.internal.pageSize.getHeight();
    const logo = await loadLogo();

    const amountColW = TABLE_COLS[5];
    const rateColW = TABLE_COLS[4];
    const totalsW = rateColW + amountColW;
    const totalsX = margin + CONTENT_W - totalsW;
    const amountX = margin + CONTENT_W - amountColW;

    drawHeader(doc, bill, pageW, margin, logo);

    let y = HEADER_H + 8;
    const gap = 5;
    const boxW = (CONTENT_W - gap * 2) / 3;
    const boxH = 28;

    drawInfoBox(
      doc,
      margin,
      y,
      boxW,
      boxH,
      "Bill To",
      bill.customerName || "Customer",
      bill.customerPhone || "—"
    );
    drawInfoBox(doc, margin + boxW + gap, y, boxW, boxH, "Delivery", formatDeliverySchedule(bill));
    drawInfoBox(
      doc,
      margin + (boxW + gap) * 2,
      y,
      boxW,
      boxH,
      "Payment",
      formatPaymentSummary(bill)
    );

    y += boxH + 10;

    const tableBody = (bill.items || []).map((item, i) => {
      const amount = (item.rate || 0) * (item.qty || 0);
      return [
        String(i + 1),
        item.name || "—",
        formatServiceName(item.service),
        formatQtyDisplay(item),
        formatCurrency(item.rate) + (isKgItem(item) ? "/kg" : ""),
        formatCurrency(amount),
      ];
    });

    doc.autoTable({
      startY: y,
      head: [["#", "Item", "Service", "Qty", "Rate", "Amount"]],
      body: tableBody,
      theme: "grid",
      margin: { left: margin, right: margin },
      tableWidth: CONTENT_W,
      headStyles: {
        fillColor: ORANGE,
        textColor: [255, 255, 255],
        fontStyle: "bold",
        fontSize: 8.5,
        cellPadding: { top: 3, right: 2, bottom: 3, left: 2 },
        halign: "center",
      },
      bodyStyles: {
        fontSize: 9,
        textColor: BLACK,
        cellPadding: { top: 2.5, right: 2, bottom: 2.5, left: 2 },
        lineColor: BORDER,
        lineWidth: 0.2,
      },
      alternateRowStyles: { fillColor: LIGHT },
      columnStyles: {
        0: { cellWidth: TABLE_COLS[0], halign: "center" },
        1: { cellWidth: TABLE_COLS[1], halign: "left" },
        2: { cellWidth: TABLE_COLS[2], halign: "left" },
        3: { cellWidth: TABLE_COLS[3], halign: "center" },
        4: { cellWidth: TABLE_COLS[4], halign: "right" },
        5: { cellWidth: TABLE_COLS[5], halign: "right", fontStyle: "bold" },
      },
    });

    y = doc.lastAutoTable.finalY + 6;

    const subtotal = bill.subtotal ?? 0;
    const discount = bill.discountAmount ?? 0;
    const discountPct = bill.discountPercent ?? 0;
    const total = bill.total ?? 0;

    doc.setFont("helvetica", "normal");
    doc.setFontSize(9);

    if (discount > 0) {
      doc.setTextColor(...GRAY);
      doc.text("Subtotal", totalsX + rateColW, y, { align: "right" });
      doc.text(formatCurrency(subtotal), margin + CONTENT_W, y, { align: "right" });
      y += 7;
      doc.setTextColor(...ORANGE);
      doc.text(`Discount (${discountPct}%)`, totalsX + rateColW, y, { align: "right" });
      doc.text(`- ${formatCurrency(discount)}`, margin + CONTENT_W, y, { align: "right" });
      y += 9;
    }

    doc.setFillColor(...ORANGE);
    doc.rect(totalsX, y, totalsW, 11, "F");
    doc.setTextColor(255, 255, 255);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(11);
    doc.text("TOTAL", totalsX + 4, y + 7.5);
    doc.text(formatCurrency(total), margin + CONTENT_W - 2, y + 7.5, { align: "right" });

    const footerY = pageH - FOOTER_H;
    const qrReserve = QR_SIZE + 10;
    const textW = pageW - margin * 2 - qrReserve;

    doc.setDrawColor(...BORDER);
    doc.setLineWidth(0.2);
    doc.line(margin, footerY, pageW - margin, footerY);

    let lineY = footerY + 6;
    doc.setTextColor(...BLACK);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(9);
    doc.text("Thank you for choosing Rinse & Rise Laundryrite!", margin + textW / 2, lineY, {
      align: "center",
    });

    lineY += 6;
    doc.setFont("helvetica", "normal");
    doc.setFontSize(8);
    doc.setTextColor(...GRAY);
    doc.text(SHOP.tagline, margin + textW / 2, lineY, { align: "center" });
    lineY += 5;
    doc.text(`Call: ${SHOP.phone}  |  ${SHOP.hours}`, margin + textW / 2, lineY, { align: "center" });
    lineY += 5;
    for (const line of SHOP.addressLines) {
      doc.text(line, margin + textW / 2, lineY, { align: "center" });
      lineY += 4;
    }

    const reviewQr = await loadReviewQr();
    if (reviewQr) {
      const qrX = pageW - margin - QR_SIZE;
      const qrY = footerY + (FOOTER_H - QR_SIZE - 8) / 2;
      doc.addImage(reviewQr, "JPEG", qrX, qrY, QR_SIZE, QR_SIZE);
      doc.setFont("helvetica", "bold");
      doc.setFontSize(7);
      doc.setTextColor(...ORANGE);
      doc.text("Rate us on Google", qrX + QR_SIZE / 2, qrY + QR_SIZE + 3, { align: "center" });
      doc.setFont("helvetica", "normal");
      doc.setFontSize(6);
      doc.setTextColor(...GRAY);
      doc.text("Scan to review", qrX + QR_SIZE / 2, qrY + QR_SIZE + 6.5, { align: "center" });
    }

    const filename = `RinseRise-Invoice-${bill.billNo || "bill"}.pdf`;
    const blob = doc.output("blob");
    return { blob, filename };
  }

  function triggerDownload(blob, filename) {
    const url = URL.createObjectURL(blob);
    const anchor = document.createElement("a");
    anchor.href = url;
    anchor.download = filename;
    anchor.style.display = "none";
    document.body.appendChild(anchor);
    anchor.click();
    document.body.removeChild(anchor);
    setTimeout(() => URL.revokeObjectURL(url), 2000);
  }

  return { generate, triggerDownload, formatCurrency, formatBillDate };
})();
